drivers/input/input_mcux_kpp.c

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright 2025, NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/input/input.h>
+#include <zephyr/logging/log.h>
+#include <fsl_kpp.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/dt-bindings/clock/imx_ccm.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/sys/util.h>
+
+LOG_MODULE_REGISTER(kpp, CONFIG_INPUT_LOG_LEVEL);
+
+#define DT_DRV_COMPAT nxp_mcux_kpp
+
+#define INPUT_KPP_COLUMNNUM_MAX  KPP_KEYPAD_COLUMNNUM_MAX
+#define INPUT_KPP_ROWNUM_MAX     KPP_KEYPAD_ROWNUM_MAX
+
+struct kpp_config {
+	KPP_Type *base;
+	const struct device *ccm_dev;
+	clock_control_subsys_t clk_sub_sys;
+	const struct pinctrl_dev_config *pcfg;
+};
+
+struct kpp_data {
+	uint32_t clock_rate;
+	struct k_work_delayable work;
+	uint8_t read_keys_old[KPP_KEYPAD_COLUMNNUM_MAX];
+	uint8_t read_keys_new[KPP_KEYPAD_COLUMNNUM_MAX];
+	uint8_t key_pressed_number;
+	const struct device *dev;
+};
+
+static void get_source_clk_rate(const struct device *dev, uint32_t *clk_rate)
+{
+	const struct kpp_config *dev_cfg = dev->config;
+	const struct device *ccm_dev = dev_cfg->ccm_dev;
+	clock_control_subsys_t clk_sub_sys = dev_cfg->clk_sub_sys;
+
+	if (!device_is_ready(ccm_dev)) {
+		LOG_ERR("CCM driver is not installed");
+		*clk_rate = 0;
+		return;
+	}
+
+	clock_control_get_rate(ccm_dev, clk_sub_sys, clk_rate);
+}
+
+static void input_kpp_data_analyze(const struct device *dev)
+{
+	const struct kpp_config *config = dev->config;
+	struct kpp_data *drv_data = dev->data;
+	uint8_t read_keys_new[KPP_KEYPAD_COLUMNNUM_MAX];
+	uint8_t read_keys_old[KPP_KEYPAD_COLUMNNUM_MAX];
```

Comment:
can't you use `read_keys_old`  straigth from the data struct? and `read_keys_new`, I don't think you need to allocate it statically, fold everything into `kpp_work_handler` and you can just have it on the stack and let it go after you are done with it